azure/pipelines/deploy/container-app.yml

```diff
@@ -0,0 +1,147 @@
+          Write-Host "##vso[task.setVariable variable=skipContainerApp]true"
+      }
+      Else {
+          Write-Host "Preconditions have been met to continue container app deployment."
+          Write-Host "##vso[task.setVariable variable=skipContainerApp]false"
+      }
+
+- ${{ if ne(variables['skipContainerApp'], 'true') }}:
+  - ${{ each step in parameters.preDeploySteps }}:
+    - ${{ step }} 
+ 
+- task: AzureCLI@2
+  displayName: '${{ coalesce(parameters.label, parameters.module) }} - Pin image'
+  inputs:
+    azureSubscription: ${{ parameters.azureRegistrySubscription }} # Container registry is typically only available in production subscription.
+    scriptType: 'pscore'
+    scriptLocation: 'inlineScript'
+    inlineScript: |
+      $container = Get-Content "${{ parameters.containerJsonPath }}" | ConvertFrom-Json
+      $registry = $($container.registry)
+      $repository = $($container.repository)
+      $image = "$($repository):$($container.version)"
+      
+      # Allow implicit Azure CLI extension install (container app).
+      az config set extension.use_dynamic_install=yes_without_prompt
+      
+      # Tag image with a environment specific name so we keep track of the image that is deployed.
+      $pinnedImage = "$image-${{ parameters.environment }}"
+      Write-Host "##vso[task.setvariable variable=pinnedImage]$pinnedImage"
+     
+      Write-Host "Tagging $image as $pinnedImage"
+      az acr import --name $registry --source "$registry/$image" --image $pinnedImage --force
+
+      # Disallow deployed image modification/deletion so that it will not automatically cleaned up by purge tasks.
+      # In other words, lock the image.
+      Write-Host "Disallowing delete of $pinnedImage"
+      az acr repository update --name $registry --image $pinnedImage --write-enabled false --delete-enabled false
+      
+      # Gather all tags that are relevant to the deployed environment.
+      # This means all images/tags that are currently deployed and the one we're about to deploy.
+      $revisions = az containerapp revision list --name ${{ parameters.containerApp }} --resource-group $resourceGroup | ConvertFrom-JSON
+      $currentTags = @( $revisions | % { $_.properties.template.containers } | % { $_.image.Split(':')[-1] } )
+      $currentTags += $pinnedImage.Split(':')[-1]
+
+      # Remove all tags for this environment that are NOT relevant anymore.
+      $allTags = az acr repository show-tags --name $registry --repository $repository | ConvertFrom-JSON
+      $tagsToRemove = $allTags | Where-Object { $_.EndsWith('-${{ parameters.environment }}') } | Where-Object { $_ -notin $currentTags }
+      foreach ($tag in $tagsToRemove) {
+        $imageToRemove = "$($repository):$tag"
+        Write-Host "Removing tag: $imageToRemove"
+
+        # Unlock tag.
+        az acr repository update --name $registry --image $imageToRemove --write-enabled true --delete-enabled true
```

Comment:
Just to be sure: Are _write-enabled_ and _delete-enabled_ attributes of the tag or the image?

In other words: Image x is/was deployed on dev and qa. It has 3 tags x, x-dev and x-qa.
We want to deploy image y (a new version) to dev. 
Image x is no longer deployed on dev, but is deployed on qa.
Tag x-dev is removed from image x and _write-enabled_ and _delete-enabled_ are reset.
Is tag x-qa still locked or not?

Reply:
Yes x-qa is still locked (have tested that manually).
The lock is on the tag.